common/primitive/atomic_u128.rs

```diff
@@ -0,0 +1,168 @@
+        (lw & SYNC_MASK) < ((uw_sync as u64) & SYNC_MASK)
+    }
+
+    #[inline]
+    fn compose(uw_sync: u128, lw: u64) -> u128 {
+        (uw_sync & UW_MASK) | (lw as u128)
+    }
+
+    pub fn get(&self) -> u128 {
+        let uw_read = self.uw_sync;
+        let lw_read  = self.lw.load(SeqCst);
+        let uw_increment = if Self::uw_outdated(uw_read, lw_read) { UW_INCREMENT } else { 0 };
+        Self::compose(uw_read + uw_increment, lw_read)
+    }
+
+    pub fn add_and_get(&mut self, increment: u32) -> u128 {
+        let uw_read = self.uw_sync;
+        let lw_before_add = self.lw.fetch_add(increment as u64, SeqCst);
+        let lw_after_add = u64::wrapping_add(lw_before_add, increment as u64);
+        let uw_increment = if Self::uw_outdated(uw_read, lw_after_add) { UW_INCREMENT } else { 0 };
+        let updated_sync_bits: bool = (lw_after_add & SYNC_MASK) != (lw_before_add & SYNC_MASK);
+
+        if updated_sync_bits {
+            self.uw_sync = ((uw_read & UW_MASK) + uw_increment) | (lw_after_add as u128);
+            Self::compose(self.uw_sync, lw_after_add)
+        } else {
+            Self::compose(uw_read + uw_increment, lw_after_add)
+        }
+    }
+
+    pub fn compare_and_exchange_incremented(&mut self, current : u128, increment: u32) -> Result<u128, u128> {
+        // UW must match to do a swap.
+        let uw_read = self.uw_sync;
+        // we can use current's LW in place of self.lw because current's sync-bits must be equal to lw's sync bits for the swap to succeed.
+        let uw_increment = if Self::uw_outdated(uw_read, current as u64) { UW_INCREMENT } else { 0 };
+        let uw_match_if_lw_match = ((uw_read & UW_MASK) + uw_increment) == (current & UW_MASK);
+
+        // Can `uw_match_if_lw_match` be true if lw(current) = lw(self) but uw(current) != uw(self) ? No.
+        //      Trying to show this should contradict that self.uw_sync cannot be ahead of self.lw.
+        // Can `uw_match_if_lw_match` be false, but the true value be equal to current? Also, No.
+        //      Sketch: Assume lw match, full match iff uw_match_if_lw_match is true.
+
+        if uw_match_if_lw_match {
+            let updated: u128 = current + increment as u128;
+            let lw_cas_result = self.lw.compare_exchange(current as u64, updated as u64, SeqCst, SeqCst);
+            match lw_cas_result {
+                Ok(lw_swapped) => {
+                    debug_assert_eq!(lw_swapped, current as u64);
+                    Ok(current)
+                },
+                Err(lw_swapped) => {
+                    let uw_increment = if Self::uw_outdated(uw_read, lw_swapped) { UW_INCREMENT } else { 0 };
+                    Err(Self::compose(uw_read + uw_increment, lw_swapped))
+                }
+            }
+        } else {
+            // If uw don't match, attempting a swap would write the wrong state to lw
+            Err(self.get())
+        }
+    }
+}
+
+pub mod tests {
```

Comment:
```suggestion
#[cfg(test)]
mod test {
```
Only needs to be compiled when testing, and doesn't need to be public.